staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -143,107 +136,168 @@ func (h *peerProxyHandler) WrapHandler(handler http.Handler) http.Handler {
+				if h.localDiscoveryCacheTicker != nil {
+					h.localDiscoveryCacheTicker.Stop()
+					klog.Info("localDiscoveryCacheTicker stopped")
+				}
+				klog.Info("localDiscoveryCacheInvalidation goroutine exiting")
+				return
+			}
 		}
+	}()
+}
 
-		if len(peerEndpoints) == 0 {
-			klog.Errorf("gvr %v is not served by anything in this cluster", gvr)
-			handler.ServeHTTP(w, r)
-			return
+func (h *peerProxyHandler) populateLocalDiscoveryCache() error {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(h.loopbackClientConfig)
+	if err != nil {
+		return fmt.Errorf("error creating discovery client: %w", err)
+	}
+
+	_, resourcesByGV, _, err := discoveryClient.GroupsAndMaybeResources()
+	if err != nil {
+		return fmt.Errorf("error getting API group resources from discovery: %w", err)
+	}
+
+	freshLocalDiscoveryResponse := map[schema.GroupVersion][]string{}
+	for gv, resources := range resourcesByGV {
+		if gv.Group == "" {
+			gv.Group = "core"
 		}
+		freshLocalDiscoveryResponse[gv] = []string{}
+		for _, resource := range resources.APIResources {
+			freshLocalDiscoveryResponse[gv] = append(freshLocalDiscoveryResponse[gv], resource.Name)
+		}
+	}
 
-		// otherwise, randomly select an apiserver and proxy request to it
-		rand := rand.Intn(len(peerEndpoints))
-		destServerHostPort := peerEndpoints[rand]
-		h.proxyRequestToDestinationAPIServer(r, w, destServerHostPort)
-	})
+	h.localDiscoveryResponseCache = freshLocalDiscoveryResponse
+	return nil
 }
 
-func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource) (*sync.Map, error) {
-	apiserversi, ok := h.svMap.Load(gvr)
-	if !ok || apiserversi == nil {
-		return nil, fmt.Errorf("no storageVersions found for the GVR: %v", gvr)
+func (h *peerProxyHandler) shouldServeLocally(gvr schema.GroupVersionResource) bool {
+	resources, ok := h.localDiscoveryResponseCache[gvr.GroupVersion()]
+	if !ok {
+		klog.V(4).Infof("resource not found for %v in local discovery cache\n", gvr.GroupVersion())
+		return false
 	}
 
-	apiservers, _ := apiserversi.(*sync.Map)
-	return apiservers, nil
+	for _, resourceName := range resources {
+		if gvr.GroupResource().Resource == resourceName {
+			return true
+		}
+	}
+
+	return false
 }
 
-func (h *peerProxyHandler) resolveServingLocation(apiservers *sync.Map) (bool, []string, error) {
-	var peerServerEndpoints []string
-	var locallyServiceable bool
-	var respErr error
-
-	apiservers.Range(func(key, value interface{}) bool {
-		apiserverKey := key.(string)
-		if apiserverKey == h.serverId {
-			locallyServiceable = true
-			// stop iteration and reset any errors encountered so far.
-			respErr = nil
-			return false
+func (h *peerProxyHandler) findServiceableByPeerFromAggDiscoveryCache(gvr schema.GroupVersionResource) []string {
+	var serviceableByIDs []string
+	var foundResource bool
+
+	h.peerAggDiscoveryCacheLock.RLock()
+	defer h.peerAggDiscoveryCacheLock.RUnlock()
+	for peerID, discoveryDoc := range h.peerAggDiscoveryResponseCache {
```

Comment:
nit: naming for discoveryDoc. Would prefer discoveryInfo or peerDiscoveryInfo

Reply:
done